### PR TITLE
JAMES-2623 port Runnables to Reactor to avoid using Common ForkJoin pool

### PR DESCRIPTION
**Pull request by mbaechler**

### CODE REVIEW

File: server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
+package org.apache.james.util;
```
Comment (chibenwa) [inline server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: licence

File: server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (chibenwa) [inline server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: wilcard

**Comment by mbaechler** [comment]
test this please

**Comment by mbaechler** [comment]
test this please

File: server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import com.google.common.base.Stopwatch;
+import reactor.core.publisher.Flux;
+
+class RunnablesTest {
+
+    @Test
+    void shouldActuallyRunThings() {
+        AtomicBoolean sideEffect = new AtomicBoolean(true);
+        Runnables.runParallel(() -> sideEffect.set(false));
+        assertThat(sideEffect).isFalse();
```
Comment (chibenwa) [inline server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: I would rather turn the boolean into true. Then it reads like 'yes, there was a side effect', which I find easier.

File: server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Test
+    void shouldActuallyRunInParallel() {
+        Runnable runnable = () -> {
+            try {
+                Thread.sleep(1000);
```
Comment (chibenwa) [inline server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: Or.... Use a countDownLatch for not being waiting for a fixed amount of time?

Comment (chibenwa) [reply server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: BTW a ThrowingRunnable would suppress the try/catch block

File: server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java (hunk 3 above)
Comment (mbaechler) [reply server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: ok

File: server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java (hunk 1 above)
Comment (chibenwa) [reply server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: Still no license

File: server/container/util/src/main/java/org/apache/james/util/Runnables.java [hunk 5]
```diff
@@ -19,27 +19,26 @@
...
-    private static Supplier<Void> toVoidSupplier(Runnable runnable) {
-        return () -> {
-            runnable.run();
-            return null;
-        };
+    public static void runParallel(Flux<Runnable> runnables, int parallel) {
+        runnables
+            .publishOn(Schedulers.elastic())
+            .parallel(parallel)
+            .runOn(Schedulers.elastic())
```
Comment (chibenwa) [inline server/container/util/src/main/java/org/apache/james/util/Runnables.java]: publishOn then runOn and 2 times on the elastic scheduler.

Could you explain the subtility of this code? Or could we simplify it?

Comment (trantienduchn) [reply server/container/util/src/main/java/org/apache/james/util/Runnables.java]: as I understand `.runOn(...)` overrides execution context of `.publishOn(...)`. So, We could remove one of two calls above

File: server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java (hunk 1 above)
Comment (mbaechler) [reply server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: I don't understand how it happened because I added it ... now fixed

Comment (chibenwa) [reply server/container/util/src/test/java/org/apache/james/util/RunnablesTest.java]: héhé you know git sometime is tricky ^^

**Comment by mbaechler** [comment]
test this please

File: server/container/util/src/main/java/org/apache/james/util/Runnables.java [hunk 6]
```diff
@@ -19,27 +19,26 @@
...
+            .flatMap(runnable -> {
+                runnable.run();
+                return Mono.empty();
+            })
+            .sequential()
+            .blockLast();
```
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/Runnables.java]: could be `.then().block()`

Comment (mbaechler) [reply server/container/util/src/main/java/org/apache/james/util/Runnables.java]: `.then().block()` make it clear we don't expect any return value while `.blockLast()` produce one that we discard

**Comment by mbaechler** [comment]
```
[3b11c05e890d54b4966f12f8045a99f1ad7d340a] [ERROR] Errors: 
[3b11c05e890d54b4966f12f8045a99f1ad7d340a] unWorker(Thre[ERROR]   ForwardSmtpTest.forwardingAnEmailShouldWork » IllegalArgument parallelism > 0 ...
[3b11c05e890d54b4966f12f8045a99f1ad7d340a] [ERROR]   SmtpStarttlsCommandTest.shouldNotRejectContentStartsWithStartTls » Provision U...
[3b11c05e890d54b4966f12f8045a99f1ad7d340a] [ERROR]   SmtpStarttlsCommandTest.shouldNotRejectContentWithStartTls » Provision Unable ...
[3b11c05e890d54b4966f12f8045a99f1ad7d340a] [ERROR]   SmtpStarttlsCommandTest.shouldNotRejectRcptWithStartTls » Provision Unable to ...
[3b11c05e890d54b4966f12f8045a99f1ad7d340a] [ERROR]   SmtpStarttlsCommandTest.starttlsShouldBeRejectedWhenFollowedByCommand » Provision
[3b11c05e890d54b4966f12f8045a99f1ad7d340a] [ERROR]   SmtpStarttlsCommandTest.starttlsShouldWork » Provision Unable to provision, se...
[3b11c05e890d54b4966f12f8045a99f1ad7d340a] [INFO] 
```

I'll have a look

File: server/container/util/src/main/java/org/apache/james/util/Runnables.java [hunk 7]
```diff
@@ -19,27 +19,26 @@
...
-                .map(runnable -> CompletableFuture.supplyAsync(toVoidSupplier(runnable))))
-            .join();
-    }
-
-    private static Supplier<Void> toVoidSupplier(Runnable runnable) {
-        return () -> {
-            runnable.run();
-            return null;
-        };
+    public static void runParallel(Flux<Runnable> runnables, int parallel) {
```
Comment (mbaechler) [inline server/container/util/src/main/java/org/apache/james/util/Runnables.java]: I'm not happy with having to give an explicit value for parallelism, I will remove that argument and rely on a default value

File: server/container/util/src/main/java/org/apache/james/util/Runnables.java (hunk 5 above)
Comment (chibenwa) [reply server/container/util/src/main/java/org/apache/james/util/Runnables.java]: Thanks for the information, it makes sens.

**Comment by rouazana** [comment]
merged

**Pull request closed without merging**